warehouse/integrations/vulnerabilities/utils.py

```diff
@@ -0,0 +1,147 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from sqlalchemy import func
+from sqlalchemy.orm.exc import NoResultFound
+
+from warehouse.integrations import vulnerabilities
+from warehouse.integrations.vulnerabilities.models import VulnerabilityRecord
+from warehouse.packaging.models import Project, Release
+
+
+def _get_project(request, vuln_report: vulnerabilities.VulnerabilityReportRequest):
+    return (
+        request.db.query(Project)
+        .filter(
+            Project.normalized_name == func.normalize_pep426_name(vuln_report.project)
+        )
+        .one()
+    )
+
+
+def _get_release(request, project: Project, version):
+    return (
+        request.db.query(Release)
+        .filter(Release.project_id == project.id)
+        .filter(Release.canonical_version == version)
+        .one()
+    )
+
+
+def _get_vuln_record(
+    request, vuln_report: vulnerabilities.VulnerabilityReportRequest, origin
+):
+    return (
+        request.db.query(VulnerabilityRecord)
+        .filter(VulnerabilityRecord.id == vuln_report.vulnerability_id)
+        .filter(VulnerabilityRecord.source == origin)
+        .one()
+    )
+
+
+def _delete_vuln_record(request, vuln_record: VulnerabilityRecord):
+    request.db.delete(vuln_record)
+
+
+def _add_vuln_record(request, vuln_record: VulnerabilityRecord):
+    request.db.add(vuln_record)
+
+
+def _analyze_vulnerability(request, vulnerability_report, origin, metrics):
+    try:
+        report = vulnerabilities.VulnerabilityReportRequest.from_api_request(
+            request=vulnerability_report
+        )
+    except vulnerabilities.InvalidVulnerabilityReportRequest as exc:
+        metrics.increment(f"warehouse.vulnerabilities.{origin}.error.{exc.reason}")
```

Comment:
general feedback on metrics calls: use `tags=[f"origin:{origin}"]` rather than embedding them in the metric name.

Reply:
(That's my bad, I did this in the PR this is inspired from. We can fix either mine or both in a subsequent PR, I guess. Do you have more general rules how metrics should be used ?)

Reply:
I would still like to see tagging implemented before merge. As this is the initial implementation and we are likely to see subsequent ones, they should be able to reference this as a best practice.

We should patch up the GitHub metrics in a separate PR @ewjoachim

Reply:
done